### ISSUE TITLE
How to get an index of currently visible top item?

### ISSUE THREAD

**Issue by mnasyrov** [rintoj/ngx-virtual-scroller#209]
Is there a way to get an index of a current item which is at the top of viewport?
As I understand "start", "end"  and "change" events provide indexes of a slice of items which is bigger than a visible viewport.
It would be nice to have events for visible items like "visibleStart" or "visibleEnd".

**Comment by speige** [comment]
Start would be the top visible item unless you have unequalChildItems or
bufferAmount set. For single column you can do start - bufferAmount, for
multiColumn it'd be start - bufferAmount*numColumns. However, as you scroll
down the start won't change until the last pixel of the top item
disappears, so it can seem like it's slightly off.

Maybe we should change this behavior to be more intuitive? I'm wondering if
we compare pixel positions & see which rendered item's center is closest to
the top & bottom.

**Comment by mnasyrov** [comment]
Yes, `enableUnequalChildrenSizes` is the my case.

**Comment by mnasyrov** [comment]
I've no time right now may be later.

**Comment by speige** [comment]
I've committed a fix that improves this, however it's still not perfect. BufferAmount is no longer included in the start/end/change events (or viewportIndices property). However, the indexes reported are sometimes less than what's actually visible.

I believe this is due to a miscalculations with enableUnequalChildrenSizes and we were hardcoding bufferAmount=5 as a workaround.

To demonstrate the issue, if I change this:
```
	public get bufferAmount(): number {
		return Math.max(this._bufferAmount, this.enableUnequalChildrenSizes ? 5 : 0);
	}
```
To this:
```
	public get bufferAmount(): number {
		return this._bufferAmount;
	}
```

And run the demo app and click "Toggle random height", the viewport indices are sometimes slightly off. Also, there is sometimes extra whitespace where there should be rendered child items.

Example (Notice it should say "Showing 15-18" and there is room for 19 & 20 to also be rendered):
![image](https://user-images.githubusercontent.com/2002030/44868011-b211cf80-ac46-11e8-8944-2b0b4ff75a08.png)
This doesn't happen in every scroll position, just some of the time.

@kykint Any ideas?

My initial guess for the extra whitespace is that as you're scrolling down, the app is using the "estimated" child measurement for any item that hasn't been rendered yet. If a new item's measured size is significantly smaller than the "estimated" size, it leaves extra room. However, my guess is probably wrong because the issue still occurs while scrolling up. Also, since refresh_internal does maxRunTimes = 2, I'd expect the 2nd run to notice the discrepancy between estimated vs actual child size measurements and fill the empty space.

I don't have any guess as to why the startIndex is sometimes smaller than it should be. I think this sometimes happens even without enableUnequalChildrenSizes when the item has only been scrolled a few pixels out of the viewport, maybe due to a rounding error. However, in my example image index 14 is hundreds of pixels out of the viewport.


Also, sometimes scrolling causes flickering when the item which can fill the remaining whitespace quickly appears & disappears. Also, the "Showing # - # of 1000 (4 nodes)" label will quickly change between different values. I think this might be due to refresh_internal maxRunTimes where each run the calculations are slightly different, causing the indexes/visible items to change during each iteration. Again, this doesn't happen every time I scroll, just some of the time.

**Comment by speige** [comment]
Note: Although I've committed a partial fix, I haven't updated the documentation or published to npm. I'm going to wait on that until the buggy side-effects are corrected.

**Comment by kykint** [comment]
I think that bufferAmount should be always, because on quick scrolling with complex elements (components with difficult structure), elements could initialize slowly. That is why for my cases even without enableUnequalChildrenSizes I have to set bufferAmount=10 to prerender items before they come to view.

**Comment by speige** [comment]
That makes sense. bufferAmount allows a few invisible elements on both sides of viewport to have measured sizes instead of estimated sizes, so scrolling will be smoother.

I think the bug is that ChangeEvent reports a start/end that don't match the viewport. I'll research what's causing that. Thanks.

**Comment by speige** [comment]
@kykint I think a found a few causes for the bug.

#1) In calculateDimensions, enableUnequalChildrenSizes loops over content.children to determine sumOfVisibleMaxWidths/sumOfVisibleMaxHeights. However, due to bufferAmount, some of the children may actually  be invisible. I think we should look at getBoundingClientRect for each child to determine if they're actually visible within the viewport.

#2) For enableUnequalChildrenSizes, dimensions.wrapGroupsPerPage and dimensions.itemsPerPage and dimensions.pageCount_fractional are based on the current page, rather than for all pages. So, we should be careful how we use them. This affects scrollToIndex_internal & calculatePageInfo

**Comment by speige** [comment]
I've committed a fix for this. It should be available on npm as version 0.4.12 soon.

**Comment by speige** [comment]
@kykint I committed a fix for this. Check it out and tell me what you think.

I've tested it & the values from change/start/end events seem more accurate (not messed up by BufferAmount). They are sometimes still off by 1, but that also happens even without enableUnequalChildrenSizes.

Also, for testing only, I made this change:
```
	public get bufferAmount(): number {
		//return Math.max(this._bufferAmount, this.enableUnequalChildrenSizes ? 5 : 0);
		return this._bufferAmount;
	}
```
and there was no extra whitespace & everything worked fine. Maybe we can consider removing BufferAmount? Can you test with your app & see if 10 is still necessary?

**Comment by Aphax** [comment]
In my case even with enableUnequalChildrenSizes set to false and bufferAmount to 0, the number of viewportItems is slightly bigger than what's displayed (17 but 13 visibles), so there is a gap of 2 rows that a pre displayed on top and bottom even with a 0 bufferAmount, is that the wished behavior ?
It would be nice to have something like an array of rendererItems with vsUpdate, what do you think ?

**Comment by speige** [comment]
1 item would be understandable because even if only 1px of the item will be visible, it needs to render it.
However, 2 extra invisible items seems like a math bug. I've just checked https://rintoj.github.io & was easily able to get 2 extra items, so it's probably common & happening for a lot of users.

I'm surprised to hear you're getting 4 extra items. I've never seen that nless enableUnequalChildrenSizes or bufferAmount are set. Which, in your case they're not.

Are all of your child items the same size?

Could you create a stackblitz to help troubleshoot your issue?

**Comment by Aphax** [comment]
@speige my bad after rechecking the bug seems to be only at bottom, like you just observe. 
There are always 2 extra items completely outside of the scroller, still confusing and not what we might expect I guess.

**Comment by speige** [comment]
I agree, this is a bug. 

I've create a separate issue for it because it's slightly different than this original issue.
#256

I don't personally have time to troubleshoot this right now. Hopefully someone else will fix it & submit a pull request :)

**Comment by speige** [comment]
@Aphax I found some time. This has been fixed & released with npm version 1.0.7. Thanks for bringing it to my attention.

**Comment by Aphax** [comment]
@speige perfect man, you rock !
